packages/test/test-end-to-end-tests/src/test/waitForSummary.spec.ts

```diff
@@ -0,0 +1,220 @@
+				errorMsg: "Timeout on waiting for handleGet",
+			});
+			assert.strictEqual(bufferToString(handleGet, "utf8"), "blob contents");
+
+			// wait for summary with redirect table
+			await timeoutAwait(provider.ensureSynchronized(), {
+				errorMsg: "Timeout on waiting for ensureSynchronized",
+			});
+			await timeoutAwait(waitForSummary(provider, container, testContainerConfig), {
+				errorMsg: "Timeout on waiting for summary",
+			});
+
+			// should be able to load entirely offline
+			const stashBlob = await timeoutAwait(
+				getPendingOps(testContainerConfig, provider, true),
+				{
+					errorMsg: "Timeout on waiting for stashBlob",
+				},
+			);
+			await timeoutAwait(loadOffline(testContainerConfig, provider, { url }, stashBlob), {
+				errorMsg: "Timeout on waiting for loadOffline",
+			});
+		});
+
+		it("applies stashed ops with no saved ops", async function () {
+			// We want to test the case where we stash ops based on the sequence number of the snapshot we load from
+			// So step 1 is to complete a summary so we can load from it.
+			const summaryVersion = await waitForSummary(provider, container, testContainerConfig);
+
+			// avoid our join op being saved (so saved ops is empty and the map op below has the right ref seq)
+			const headers: IRequestHeader = {
+				[LoaderHeader.loadMode]: { deltaConnection: "none" },
+				[LoaderHeader.version]: summaryVersion,
+			};
+			const container2: IContainerExperimental = await loader.resolve({ url, headers });
+			const dataStore2 = (await container2.getEntryPoint()) as ITestFluidObject;
+			const map2 = await dataStore2.getSharedObject<ISharedMap>(mapId);
+			// generate ops with RSN === summary SN
+			map2.set("2", "2");
+			const stashBlob = await container2.closeAndGetPendingLocalState?.();
+			assert(stashBlob);
+			const pendingState = JSON.parse(stashBlob);
+
+			// make sure the container loaded from summary and we have no saved ops
+			assert.strictEqual(pendingState.savedOps.length, 0, "Expected no saved ops");
+			assert(
+				pendingState.pendingRuntimeState.pending.pendingStates[0].referenceSequenceNumber > 0,
+				"Expected the pending state to have some ops with non-zero ref seq (should match the snapshot sequence number)",
+			);
+
+			// load container with pending ops, which should resend the op not sent by previous container
+			const container3 = await loader.resolve({ url }, stashBlob);
+			const dataStore3 = (await container3.getEntryPoint()) as ITestFluidObject;
+			const map3 = await dataStore3.getSharedObject<ISharedMap>(mapId);
+			await waitForContainerConnection(container3);
+			await provider.ensureSynchronized();
+			assert.strictEqual(map1.get("2"), "2");
+			assert.strictEqual(map3.get("2"), "2");
+		});
+
+		it("can stash between summary op and ack", async function () {
+			const waitForSummaryPromise = waitForSummary(provider, container, testContainerConfig);
+			const pendingOps = await new Promise<string | undefined>((resolve, reject) =>
```

Comment:
Use `timeoutPromise` or `timeoutAwait` here in case the summarize op never comes.